### PR TITLE
Authentication test

### LINKED ISSUE #499: Automated UI test for authentication
Section 8 in TESTING.md

### PR DESCRIPTION
**Pull request by jaakkonarhi**
# Description of the PR

+ Closes #499 
+ Uses the token of the test user testik34

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt [hunk 1]
```diff
@@ -18,7 +18,9 @@ class PlaceholderTest {
 
   @Test
   fun mainTest() = uiTest {
+    //step 2
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:21]: nice idea 😉 (I assume these are the steps from [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md))?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt [hunk 2]
```diff
@@ -55,5 +57,25 @@ class PlaceholderTest {
...
+      }
+      with(dialog("A+ Token")) {
+        jPasswordField().text = System.getenv("APLUS_TEST_TOKEN")
+        button("OK").click()
+      }
+      with(ideFrame()) {
+        with(assignments()) {
+          waitFor(
+            Duration.ofSeconds(60),
+            Duration.ofSeconds(1),
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:74]: why not `Duration.ofSeconds(61)`?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/fixtures/CommonFixtures.kt [hunk 3]
```diff
@@ -71,6 +76,8 @@ class DialogFixture(remoteRobot: RemoteRobot, remoteComponent: RemoteComponent)
     : CommonContainerFixture(remoteRobot, remoteComponent) {
   fun ContainerFixture.sidePanel() = find(ContainerFixture::class.java,
       byXpath("//div[@class='SidePanel']"))
+  fun jPasswordField() = find(JTextFieldFixture::class.java,
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/fixtures/CommonFixtures.kt:79]: this is something I'm not sure about... `jPasswordField` feels being a very special **Swing** component, whereas others, e.g. `sidePanel` in this same `Fixture` and elsewhere are a bit more abstract. 

IMO, perhaps we could aim for some higher-level abstraction and the person writing the code should not explicitly know, that s/he's working with `jPasswordField` but rather a  more generic `passwordField`. This is mostly about the naming, the code is complete. However, I believe we could/should agree on a common approach here. 

This also concerns L47 (`aPlusStripeButton`).

**Comment by superseacat** [comment]
I'm afraid there's a build failing 😝 
Screenshot from 2021-02-24 20-14-24

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -47,4 +47,6 @@ jobs:
           Xvfb -ac :99 -screen 0 1920x1080x16 &
           gradle runIdeForUiTests &
       - name: Run the UI tests
+        env:
+          APLUS_TEST_TOKEN: ${{ secrets.APLUS_TEST_TOKEN }}
```
Comment (superseacat) [inline .github/workflows/build.yml:51]: I suggest thinking carefully about the credentials used for UI testing, as we don't have many pairs of them, and the one used in automation will get polluted very very quickly (once we get to submissions) @Aalto-LeTech/apluscourses-plugin-team

**Review by superseacat — COMMENTED** [review]
shall we have a checkstyle for **Kotlin** (does not have to be a part of this PR)?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 2 above)
Comment (jaakkonarhi) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:74]: the first duration is the timeout and the second one is how often it's checked

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 1 above)
Comment (jaakkonarhi) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:21]: yes they are

**Review by superseacat — APPROVED** [review]
good to go! 🚀

**Comment by stellatsv** [comment]
The test case that you included is good but it is not yet handled properly by the plugin. When the user sets an invalid, non-empty token there is an exception thrown #272. This exception should perhaps not be considered as an ok/acceptable outcome in the tests? Maybe the test could be merged as-is for now and changed once the issue is fixed, what do you think @superseacat? 

The test case that I described in my previous comment was to leave the token empty and press ok. The test could check that there is a warning displayed in the window.

**Comment by superseacat** [comment]
> The test case that you included is good but it is not yet handled properly by the plugin. When the user sets an invalid, non-empty token there is an exception thrown #272. This exception should perhaps not be considered as an ok/acceptable outcome in the tests? Maybe the test could be merged as-is for now and changed once the issue is fixed, what do you think @superseacat?

could it be possible to cover the test in #272 (perhaps an extra comment to remind there) and merge this once other leftovers are done?

**Review by stellatsv — APPROVED** [review]
Awesome work!

**Pull request merged**